hera_qm/firstcal_metrics.py

```diff
@@ -0,0 +1,533 @@
+            if self.ant_std[ant] > std_cut:
+                self.bad_ants.append(ant)
+
+        # put into dictionary
+        result = OrderedDict()
+        result['full_sol'] = self.full_sol
+        result['bad_ant'] = self.bad_ants
+        result['z_scores'] = self.z_scores
+        result['ant_std'] = self.ant_std
+        result['time_std'] = self.time_std
+        result['agg_std'] = self.agg_std
+        result['times'] = self.times
+        self.result = result
+
+        if output == True:
+            return self.result
+       
+    def write_metrics(self, filename=None, filetype='pkl'):
+        """
+        Write metrics to file after running run_metrics()
+
+        filename : str, default=None
+            filename without filetype suffix
+            if None, use default filename stem
+
+        filetype : str, default='pkl'
+            filetype to write out to
+            options = ['json', 'pkl']
+
+        """
+        # get filename prefix
+        if filename is None:
+            filename = self.file_stem + ".metrics"
+
+        # write to file
+        if filetype == 'json':
+            filename +='.json'
+            # change ndarrays to lists
+            self.result['times'] = list(self.result['times'])
+            for k in self.result['z_scores'].keys():
+                self.result['z_scores'][k] = list(self.result['z_scores'][k])
+                   
+            result_str = json.dumps(self.result)
+            with open(filename, 'w') as f:
+                json.dump(result_str, f, indent=4)
+
+        elif filetype == 'pkl':
+            filename += '.pkl'
+            with open(filename, 'wb') as f:
+                outp = pkl.Pickler(f)
+                outp.dump(self.result)
+
+    def load_metrics(self, filename):
+        """
+        Read-in a firstcal_metrics file and append to class
+
+        Input:
+        ------
+        filename : str
+            filename to read in
+
+        """
+        # get filetype
+        filetype = filename.split('.')[-1]
+        if filetype == 'json':
+            with open(filename, 'r') as f:
+                self.result = json.load(f)
+
+        elif filetype == 'pkl':
+            with open(filename, 'rb') as f:
+                inp = pkl.Unpickler(f)
+                self.result = inp.load() 
+        else:
+            raise IOError("Filetype not recognized, try a json or pkl file")
+
+
+    def delay_std(self, return_dict=False):
+        """
+        Calculate standard deviations of per-antenna delay solutions
+        and aggregate delay solutions. Assign a z-score for
+        each (antenna, time) delay solution w.r.t. aggregate
+        mean and standard deviation.
+
+        Uses astropy.stats.biweight_midvariance for a robust measure
+        of the std
+
+        Input:
+        ------
+
+        return_dict : bool, [default=False]
+            if True, return time_std, ant_std and z_scores
+            as a dictionary with "ant" as keys in str
```

Comment:
Not essential for now, but we should try to be consistent between different modules - ant_metrics uses ants as integers as keys.